sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala

```diff
@@ -88,14 +89,70 @@ case class AnalyzeTableCommand(tableName: String) extends RunnableCommand {
             }
           }.getOrElse(0L)
 
-        // Update the Hive metastore if the total size of the table is different than the size
-        // recorded in the Hive metastore.
-        // This logic is based on org.apache.hadoop.hive.ql.exec.StatsTask.aggregateStats().
-        if (newTotalSize > 0 && newTotalSize != oldTotalSize) {
+        var needUpdate = false
+        val totalSize = if (newTotalSize > 0 && newTotalSize != oldTotalSize) {
+          needUpdate = true
+          newTotalSize
+        } else {
+          oldTotalSize
+        }
+        var numRows: Option[BigInt] = None
+        if (!noscan) {
+          val oldRowCount: Long = if (catalogTable.catalogStats.isDefined) {
+            catalogTable.catalogStats.get.rowCount.map(_.toLong).getOrElse(-1L)
+          } else {
+            -1L
+          }
+          val newRowCount = sparkSession.table(tableName).count()
+          if (newRowCount >= 0 && newRowCount != oldRowCount) {
+            numRows = Some(BigInt(newRowCount))
+            needUpdate = true
+          }
+        }
+        // Update the metastore if the above statistics of the table are different from those
+        // recorded in the metastore.
+        if (needUpdate) {
+          sessionState.catalog.alterTable(
+            catalogTable.copy(
+              catalogStats = Some(Statistics(
+                sizeInBytes = totalSize, rowCount = numRows))),
+            fromAnalyze = true)
+
+          // Refresh the cache of the table in the catalog.
+          sessionState.catalog.refreshTable(tableIdent)
+        }
+
+      // data source tables have been converted into LogicalRelations
+      case logicalRel: LogicalRelation if logicalRel.metastoreTableIdentifier.isDefined =>
```

Comment:
you can run my added test case "test table-level statistics for data source table created in HiveExternalCatalog"